### PR TITLE
Add Pair module to standard library

### PR DESCRIPTION
**Pull request by lupjo**
This is related to issue #11651

This adds a module `Pair` to the standard library.

I'm not familiar with the structure of OCaml's code, so I couldn't get it to run, Stdlib.Pair is not detected it seems, therefore I cannot run the tests. I am sure the maintainers will be able to fix this just by looking at the PR, and I thank you in advance.

### CODE REVIEW

**Comment by nojb** [comment]
At the very least you will need to add the new module to `stdlib/StdlibModules`. You will also need to run `make depend` (you will need to do this in an already built tree).

**Comment by lupjo** [comment]
Thanks Nicolás, pushed fixed version

**Review by nojb — COMMENTED** [review]
- Some of the tests (which have generated id timestamps in them) need to be updated:
```
List of failed tests:
    tests/basic/patmatch_for_multiple.ml
    tests/match-side-effects/check_partial.ml
    tests/match-side-effects/partiality.ml
    tests/match-side-effects/test_contexts_code.ml
```
- This needs a `Changes` file.

**Comment by nojb** [comment]
@dbuenzli would you care to take a look at this proposal?

File: stdlib/pair.mli
Comment (dbuenzli) [inline stdlib/pair.mli]: Looking at the [other `Stdlib` synopses](https://ocaml.org/manual/5.2/api/index.html) I think this should be either `Pair values.` or `Operations on pairs`.

File: stdlib/pair.mli
Comment (dbuenzli) [inline stdlib/pair.mli:28]: I think it would be nice to add `Pair.fst` and `Pair.snd` here.

File: stdlib/pair.mli
Comment (dbuenzli) [inline stdlib/pair.mli]: Please add a label `{1:pairs Pairs}`.

File: stdlib/pair.mli
Comment (dbuenzli) [inline stdlib/pair.mli]: Please add a label `{1:preds Predicates and comparisons}`

File: stdlib/pair.mli
Comment (dbuenzli) [inline stdlib/pair.mli]: Please add a label `{1:iterators Iterators}`.

**Review by dbuenzli — CHANGES_REQUESTED** [review]
This looks mostly fine to me. 

But I have to say, some of the definitions look rather on the side of "let's have a gratuitously consistent design" rather than driven by use. By which I mean:

1. I don't see myself using some of these functions in practice. For example, will I ever use `Pair.iter_fst` ? 

    ```ocaml
    let () = Pair.iter_fst f p 
    let () = f (fst p)
    let () = f (Pair.fst p)
    let () = List.iter (Pair.iter_fst f) l 
    let () = List.iter (fun (v, _) = f v) l
    ```

2. I would likely need to lookup what `Pair.iter2 f g p0 p1` really does when I read it.

For me the iterators could have been left to either of the following sets:

```ocaml
Pair.{fold,uncurry}
Pair.{fold,uncurry,map,iter}
```
But if people like it that way I don't mind either.

**Comment by nojb** [comment]
Thanks @dbuenzli for the review! I also agree that `{map,iter}_{fst,snd}` and `{map,iter}2` could be left out of an initial version. @lupjo can you do these changes (if you agree), and fix the tests? Thanks!

**Comment by xvw** [comment]
> Thanks @dbuenzli for the review! I also agree that `{map,iter}_{fst,snd}` and `{map,iter}2` could be left out of an initial version. @lupjo can you do these changes (if you agree), and fix the tests? Thanks!

From my POV, `map_fst/snd` are very useful (and I use them a lot in personal code base)

File: stdlib/pair.mli
Comment (xvw) [reply stdlib/pair.mli]: And probably adding a `@since`

**Comment by nojb** [comment]
> > Thanks @dbuenzli for the review! I also agree that `{map,iter}_{fst,snd}` and `{map,iter}2` could be left out of an initial version. @lupjo can you do these changes (if you agree), and fix the tests? Thanks!
> 
> From my POV, `map_fst/snd` are very useful (and I use them a lot in personal code base)

Yes, sorry, we can leave `map_{fst,snd}` in, but `iter_{fst,snd}` seems less useful... do you agree?

File: stdlib/pair.mli
Comment (Octachron) [inline stdlib/pair.mli]: `Pair.uncurry` reads strangely to me. I would propose to drop the function since it is a duplicate for `fold`.

Comment (dbuenzli) [reply stdlib/pair.mli]: `curry` and `uncurry` are standard haskellisms.

I agree it should be dropped: if desirable such functions would likely make more sense in the `Fun` module.

**Comment by lupjo** [comment]
I just pushed the proposed changes, with which I agree :)
I have no idea what I did with the timestamp id, I suppose I didn't break anything, since I just changed the numbers, but if somebody wants to check that would be fine

Adding curry and uncurry to `Fun` might be a valid idea, but it's a project for a different PR I guess

**Comment by lthls** [comment]
The stamp change is expected: the `Stdlib` module contains one extra field, and it is opened by default, so all stamps after that get shifted by one. You can promote the relevant tests, and hopefully someone will double-check before merging.

**Review by dbuenzli — APPROVED** [review]
That looks good to me except for one usability quirk. 

I would like as much as possible NOT to see the name `(a, b) Pair.t` or equations that include it in my error messages. I like my regular `'a * 'b`, I don't need new names for the same thing :-) As it stands this is likely to happen as soon as I start using a function of the `Pair` module. 

As such I would advise not to use the `t` type name in the signature of any of the functions of the module (including `make`). That is replace all these `('a, 'b) t` by `'a * 'b`.

**Comment by lupjo** [comment]
Thanks @nojb for the fixes, paste and copy is a dangerous business indeed...
@dbuenzli I have no horse in this race, if no one disagrees I'm fine with using `'a * 'b` instead of `('a, 'b) t` in the signatures, I'll push the fix tomorrow
EDIT: Yeah I just realized "tomorrow" is not very useful without a timezone, so let's say "in the next 24 hours" instead ;)

**Comment by lupjo** [comment]
Also, it's consistent with `option`, `list` and `array` to make this change, which seems like a relevant point

**Comment by nojb** [comment]
> Thanks @nojb for the fixes, paste and copy is a dangerous business indeed...

No problem; I also took the liberty of fixing a remaining test (coming from recent changes in `trunk`) and implementing @dbuenzli's suggestion.

**Review by nojb — APPROVED** [review]
LGTM

As a `stdlib` addition, this PR requires a second approval from a core dev. cc possible candidates for such an approval @Octachron @gasche @yallop.

**Comment by nojb** [comment]
> LGTM
> 
> As a `stdlib` addition, this PR requires a second approval from a core dev. cc possible candidates for such an approval @Octachron @gasche @yallop.

Friendly ping; the contributor has done everything that has been asked, the ball is in our court!

File: stdlib/pair.ml
Comment (yallop) [inline stdlib/pair.ml]:
```suggestion
(*   Copyright 2024 Institut National de Recherche en Informatique et     *)
```

File: stdlib/pair.mli
Comment (yallop) [inline stdlib/pair.mli]:
```suggestion
(*   Copyright 2024 Institut National de Recherche en Informatique et     *)
```

File: stdlib/pair.mli
Comment (yallop) [inline stdlib/pair.mli]:
```suggestion
  @since 5.4 *)
```

**Review by yallop — CHANGES_REQUESTED** [review]

File: stdlib/pair.mli
Comment (lupjo) [reply stdlib/pair.mli]: Ok but why? 5.3 isn't released yet right?

Comment (yallop) [reply stdlib/pair.mli]: It's been branched, and is in feature-freeze.

Comment (lupjo) [reply stdlib/pair.mli]: Alright then ;) commited, thanks

**Review by yallop — APPROVED** [review]

**Comment by nojb** [comment]
I took the liberty of synchronizing with `trunk` and adding @yallop as a reviewer.

Planning to merge once CI passes. Thanks!

**Comment by nojb** [comment]
> Planning to merge once CI passes.

(squash-merge!)

**Comment by lupjo** [comment]
Thanks everyone 😊

**Pull request merged**
